### PR TITLE
Cloudwatch alarms

### PR DESCRIPTION
**Pull request by Vermyndax**
Part of #13.

### CODE REVIEW

File: terraform/env/cloudwatch.tf [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  comparison_operator       = "GreaterThanOrEqualToThreshold"
+  evaluation_periods        = "2"
+  period                    = "300"
+  statistic                 = "Average"
+  threshold                 = "80"
+  alarm_description         = "This metric monitors ec2 memory utilization for the instance ${aws_instance.wordpress.id}"
+  alarm_actions = ["${aws_sns_topic.sns_general_availability.arn}"]
+
+  dimensions {
+    InstanceId = "${aws_instance.wordpress.id}"
```
Comment (afeld) [inline terraform/env/cloudwatch.tf]: Unclear how to make it apply to all instances.

File: terraform/env/iam.tf [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/mon-scripts.html#using_put_examples
+module "wordpress_role" {
+  source = "github.com/18F/cg-provision/terraform/modules/iam_role"
+  iam_policy = "${file("${path.module}/files/role_policy.json")}"
+  role_name = "${var.wordpress_iam_role_name}"
+}
```
Comment (afeld) [inline terraform/env/iam.tf]: @Vermyndax I didn't put all this new code in a module since you warned about modules in modules. Open to ideas.

**Comment by afeld** [comment]
It works!

screen shot 2017-12-06 at 2 46 39 am

With an artificially lowered threshold, here's the alert sent to email:

screen shot 2017-12-06 at 2 48 03 am

File: ansible/wordpress.yml [hunk 3]
```diff
@@ -10,6 +10,12 @@
     wp_content_dir: "{{ wp_install_dir }}/wp-content"
   roles:
     - ansible-os-ubuntu-16
+    - role: cristian04.cloudwatch-monitoring
+      cron_specs:
+        - name: "Cloudwatch monitoring"
+          minute: "*/5"
```
Comment (afeld) [inline ansible/wordpress.yml]: Is this frequent enough?

File: terraform/env/cloudwatch.tf [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+  # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/mon-scripts.html#using_put_script
+  metric_name               = "MemoryUtilization"
+  namespace                 = "System/Linux"
+
+  comparison_operator       = "GreaterThanOrEqualToThreshold"
+  evaluation_periods        = "2"
+  period                    = "300"
+  statistic                 = "Average"
+  threshold                 = "80"
```
[the remark this answers is not in the document]
Comment (Vermyndax) [reply terraform/env/cloudwatch.tf]: I would suggest that we set this as a variable with a default of 80.

File: terraform/env/cloudwatch.tf (hunk 1 above)
Comment (Vermyndax) [reply terraform/env/cloudwatch.tf]: https://github.com/hashicorp/terraform/issues/8600

File: terraform/mgmt/cloudwatch.tf [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  statistic                 = "Average"
+  threshold                 = "${var.high_cpu_jenkins_alarm_threshold}"
+  alarm_description         = "This metric monitors ec2 cpu utilization and alarms for 80% over two eval periods on instance $${module.jenkins_instances.instance_id}"
+  alarm_actions = ["${aws_sns_topic.sns_general_availability.arn}"]
+
+  dimensions {
+    InstanceId = "${module.jenkins_instances.instance_id}"
+  }
+  insufficient_data_actions = []
+}
```
Comment (afeld) [inline terraform/mgmt/cloudwatch.tf]: I'd suggest keeping this pull request limited to the monitoring of the WordPress _or_ the Jenkins instance, but not both. Keeps it smaller, and we can add the other later, potentially figuring out how to share code once we have it a bit more established.

File: ansible/wordpress.yml (hunk 3 above)
Comment (Vermyndax) [reply ansible/wordpress.yml]: Anything over 5 min sampling requires detailing monitoring, which incurs costs.

**Comment by Vermyndax** [comment]
@afeld I think this is ready for merging now.

File: ansible/wordpress.yml (hunk 3 above)
Comment (afeld) [reply ansible/wordpress.yml]: Let’s add that as a comment then.

File: terraform/mgmt/jenkins.tf [hunk 6]
```diff
@@ -14,7 +14,7 @@ data "aws_ami" "jenkins" {
 
   filter {
     name = "name"
-    values = ["jenkins *"]
+    values = ["jenkins*"]
```
Comment (afeld) [inline terraform/mgmt/jenkins.tf]: Out of curiosity, why this change?

File: terraform/mgmt/sns.tf [hunk 7]
```diff
@@ -0,0 +1,3 @@
+resource "aws_sns_topic" "sns_general_availability" {
+  name = "${var.sns_general_availability_topic_name}"
+}
```
Comment (afeld) [inline terraform/mgmt/sns.tf]: Oops, this should be in the new module.

File: terraform/env/monitoring.tf [hunk 8]
```diff
@@ -0,0 +1,19 @@
+module "alerting" {
+  source = "../modules/alerting"
+
+  sns_general_availability_topic_name = "${var.sns_general_availability_topic_name}"
+  general_availability_protocol = "${var.general_availability_protocol}"
+  general_availability_endpoint = "${var.general_availability_endpoint}"
+}
```
Comment (afeld) [inline terraform/env/monitoring.tf]: Oops, I guess this should be in `mgmt`?

File: terraform/mgmt/jenkins.tf (hunk 6 above)
Comment (Vermyndax) [reply terraform/mgmt/jenkins.tf]: In my case, when I build AMI’s, I use dashes as a separator instead of spaces - like “jenkins-121117-01”. This filter won’t find them. It’s a nit.

File: terraform/env/monitoring.tf (hunk 8 above)
Comment (Vermyndax) [reply terraform/env/monitoring.tf]: In my mind, this would be part of the bootstrap environment, but could go either way. Since the SNS is a resource that isn't tied to a particular VPC and both environments use it, I thought maybe it should go in bootstrap. But it could be argued either way, that's just how I usually think. ¯\_(ツ)_/¯

Comment (afeld) [reply terraform/env/monitoring.tf]: I was keeping `bootstrap` to the bucket for the Terraform backend. Since the state is only stored locally, I wouldn't want to put more in that module. Maybe the name `bootstrap` is misleading?

File: ansible/wordpress.yml [hunk 9]
```diff
@@ -10,6 +10,13 @@
     wp_content_dir: "{{ wp_install_dir }}/wp-content"
   roles:
     - ansible-os-ubuntu-16
+    - role: cristian04.cloudwatch-monitoring
+    # CloudWatch basic monitoring uses 5 minute intervals. Enhanced monitoring supports up to 1 minute intervals, but results in extra charges.
+      cron_specs:
+        - name: "Cloudwatch monitoring"
+          minute: "*/5"
+          hour: "*"
+          job: "{{ cloud_watch_monitoring_path }}/aws-scripts-mon/mon-put-instance-data.pl --mem-util --mem-used --mem-avail --swap-util --swap-used --disk-space-util --disk-path={{ path }} --disk-space-avail --from-cron"
```
Comment (afeld) [inline ansible/wordpress.yml]: Not a blocker: Maybe we put all this in a `monitoring_cron_specs` variable under `group_vars/all/` so that it can be shared everywhere?

File: terraform/env/monitoring.tf (hunk 8 above)
Comment (afeld) [reply terraform/env/monitoring.tf]: Anyway, if we only want one of them, we just keep the one under `mgmt` and delete this block.

File: terraform/modules/alerting/main.tf [hunk 10]
```diff
@@ -0,0 +1,10 @@
+resource "aws_sns_topic" "sns_general_availability" {
+  name = "${var.sns_general_availability_topic_name}"
+
+  # https://github.com/hashicorp/terraform/issues/10697#issuecomment-312029088
+  # SNS not supported in all regions
+  # https://docs.aws.amazon.com/sns/latest/dg/sms_supported-countries.html
```
Comment (afeld) [inline terraform/modules/alerting/main.tf]: Not a blocker: meant to move these two lines to the `description` of the `general_availability_protocol` `variable`.

File: terraform/modules/os_monitoring/cloudwatch.tf [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  }
+}
+
+resource "aws_cloudwatch_metric_alarm" "high_cpu" {
+  alarm_name                = "high-cpu-${var.component}"
+  comparison_operator       = "GreaterThanOrEqualToThreshold"
+  evaluation_periods        = "2"
+  metric_name               = "CPUUtilization"
+  namespace                 = "AWS/EC2"
+  period                    = "120"
```
Comment (afeld) [inline terraform/modules/os_monitoring/cloudwatch.tf]: Not a blocker: Pretty sure I just copied-and-pasted `period` and `evaluation_periods` values here. Should we parameterize? Does it make sense to keep them different?

Comment (Vermyndax) [reply terraform/modules/os_monitoring/cloudwatch.tf]: Yeah, we should parameterize all of that, but as this branch has gotten huge, let's do another branch on that after closing this PR.
